### PR TITLE
Disable lock file if directory does not exist or no permissions

### PR DESCRIPTION
**Pull request by Ruadhri17**
## Proposed changes
This PR disable lock file mechanism when `{run.path}/lock` does not exist or directory is not writable.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#2131
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_agent/src/agent.rs [hunk 1]
```diff
@@ -101,11 +102,13 @@ pub struct Agent {
...
     pub fn try_new(name: &str, config: AgentConfig) -> Result<Self, FlockfileError> {
         let mut flock = None;
-        if config.use_lock {
+        if config.use_lock && can_create_lock_file(config.run_dir.as_std_path()) {
             flock = Some(check_another_instance_is_not_running(
                 name,
                 config.run_dir.as_std_path(),
             )?);
+        } else {
+            info!("The locking mechanism is not active");
```
Comment (PradeepKiruvale) [inline crates/core/tedge_agent/src/agent.rs:111]:
```suggestion
            info!("The locking mechanism is not active, so it does not prevent running multiple instances of tedge-agent");
```

File: crates/core/tedge_mapper/src/main.rs [hunk 2]
```diff
@@ -102,12 +104,14 @@ async fn main() -> anyhow::Result<()> {
...
     let mut _flock = None;
-    if config.run.lock_files {
+    if config.run.lock_files && can_create_lock_file(config.run.path.as_std_path()) {
         let run_dir = config.run.path.as_std_path();
         _flock = Some(check_another_instance_is_not_running(
             &mapper_opt.name.to_string(),
             run_dir,
         )?);
+    } else {
+        info!("The locking mechanism is not active");
```
Comment (PradeepKiruvale) [inline crates/core/tedge_mapper/src/main.rs:114]:
```suggestion
       info!("The locking mechanism is not active, so it does not prevent running multiple instances of tedge-mapper");
```

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
The proposed check is not working: if a user cannot write the lock while another user could, then this user will not be able to run the agent or the mapper.

A better approach is to leverage the error returned by [`Flockfile::new_lock()`](https://github.com/thin-edge/thin-edge.io/blob/3e37b354ed8937e6870ae9094667ffedeb5d414c/crates/common/flockfile/src/unix.rs#L161C18-L161C18), ignoring the error [`if !error.non_exclusive_access()`](https://github.com/thin-edge/thin-edge.io/blob/3e37b354ed8937e6870ae9094667ffedeb5d414c/crates/common/flockfile/src/unix.rs#L45)

**Comment by didier-wenzek** [comment]
> A better approach is to leverage the error returned by [`Flockfile::new_lock()`](https://github.com/thin-edge/thin-edge.io/blob/3e37b354ed8937e6870ae9094667ffedeb5d414c/crates/common/flockfile/src/unix.rs#L161C18-L161C18), ignoring the error [`if !error.non_exclusive_access()`](https://github.com/thin-edge/thin-edge.io/blob/3e37b354ed8937e6870ae9094667ffedeb5d414c/crates/common/flockfile/src/unix.rs#L45)

The fix is a bit more involved as  [`non_exclusive_access()`](https://github.com/thin-edge/thin-edge.io/blob/3e37b354ed8937e6870ae9094667ffedeb5d414c/crates/common/flockfile/src/unix.rs#L45) should check [errno](https://linux.die.net/man/3/lockf) has been set to [`EAGAIN `](https://docs.rs/nix/latest/nix/errno/enum.Errno.html#variant.EAGAIN) or [`EACCES`](https://docs.rs/nix/latest/nix/errno/enum.Errno.html#variant.EACCES).

**Review by didier-wenzek — COMMENTED** [review]
The code is correct and behaves as described by https://github.com/thin-edge/thin-edge.io/issues/2131

However, the warning message published when locks are disabled could be more useful for the user who intends to enable locks.

```
$ tedge config set run.path /tmp
$ target/debug/tedge-mapper --debug collectd 
2023-08-28T07:34:59.955514889Z  INFO flockfile::unix: The locking mechanism is not active, so it does not prevent running multiple instances of tedge-mapper-collectd.
```

This should tell the user *why* locking mechanism is not active.

```
The locking mechanism is not active, as `/tmp/lock` is not a directory with write access.
```

Might be a bit verbose though. @reubenmiller what's your point of view?

**Comment by reubenmiller** [comment]
> ```
> The locking mechanism is not active, as `/tmp/lock` is not a directory with write access.
> ```

I really like the Didier's suggestion, so please use the statement.

```
The locking mechanism is not active, as `/tmp/lock` is not a directory with write access.
```

**Pull request merged**